### PR TITLE
Fix stack overflow detection with systhreads

### PR DESCRIPTION
**Pull request by stedolan**
Stack overflow detection doesn't work properly with systhreads, as shown by #7490. A fix was proposed a while ago in #1062, but stalled. This PR implements a simpler approach.

Instead of trying to work out the stack dimensions by querying OS interfaces, this patch defines a stack overflow as a SIGSEGV that happens on an address that's between the top of the stack and the current stack pointer (or slightly past the stack pointer, to handle some edge cases around function prologues).

There are a couple of changes:

  - calls to getrlimit are removed from the signal handler, inspecting the stack pointer instead.

  - all systhreads now reinitialise the sigaltstack (but not the signal handler) on startup. ~~POSIX says you should have to do the former but not the latter, on Linux it seems you need both~~ (this is nonsense, I was testing it badly. Reinitialising sigaltstack and leaving the handler alone suffices). All threads share the same sigaltstack - they won't collide unless there's a segfault in C code, in which case all bets are already off.

  - the stack probe limit on amd64 was lowered from 32k to 4k. Linux adds a 4k guard page at the end of thread stacks to catch overflow, but beyond that is free space. It is quite common (i.e. happened regularly in testing) for that space to be used by some other allocation. So, a stack probe of 32k from the current stack pointer has false positives: it crosses the guard page and touches some random bit of memory. A 4k probe cannot have false positives: it either lies in the stack or in the guard page.

  - `caml_get_exception_raw_backtrace` no longer allocates an 8K array on the stack. (This is an independent change, in a seperate commit. Including it here makes it safe to call `Printexc.print_backtrace` from a `Stack_overflow` handler).

This has so far only been tested on Linux amd64, but I've added CONTEXT_SP support for other platforms. Hopefully it will work everywhere?

### CODE REVIEW

**Comment by stedolan** [comment]
cc @mshinwell, @xavierleroy

File: runtime/signals_nat.c [hunk 1]
```diff
@@ -185,37 +179,29 @@ DECLARE_SIGNAL_HANDLER(trap_handler)
...
-/* PR#4746: recent Linux kernels with support for stack randomization
-   silently add 2 Mb of stack space on top of RLIMIT_STACK.
-   2 Mb = 0x200000, to which we add 8 kB (=0x2000) for overshoot. */
-#define EXTRA_STACK 0x202000
-#else
-#define EXTRA_STACK 0x2000
-#endif
+/* Code compiled with ocamlopt never accesses more than
+   EXTRA_STACK bytes below the stack pointer. */
+#define EXTRA_STACK 256
```
Comment (avsm) [inline runtime/signals_nat.c:198]: Do we need to take [pthread_attr_setguardsize(3)](https://linux.die.net/man/3/pthread_attr_setguardsize) into account here? My understanding is that the thread stack guard page can be disabled by setting this to 0 if the application wants to save memory when running with many threads.

Comment (stedolan) [reply runtime/signals_nat.c:198]: I don't think so. If there's no guard page (and you're unlucky enough to have data allocated right next to a stack), then you'll never see a stack overflow segfault and this will never run.

Disabling the guard page breaks stack overflow detection, but that's basically what it's for. I don't think we need to detect and react to it in any way.

File: runtime/signals_nat.c [hunk 2]
```diff
@@ -185,37 +179,29 @@ DECLARE_SIGNAL_HANDLER(trap_handler)
...
      - faulting address is word-aligned
-     - faulting address is within the stack
+     - faulting address is on the stack, or within EXTRA_STACK of it
      - we are in OCaml code */
   fault_addr = CONTEXT_FAULTING_ADDRESS;
   if (((uintnat) fault_addr & (sizeof(intnat) - 1)) == 0
-      && getrlimit(RLIMIT_STACK, &limit) == 0
-      && fault_addr < system_stack_top
-      && fault_addr >= system_stack_top - limit.rlim_cur - EXTRA_STACK
+      && fault_addr < caml_top_of_stack
```
Comment (xavierleroy) [inline runtime/signals_nat.c]: I'm afraid `caml_top_of_stack` is valid only for the initial thread and makes no sense for other threads.

I though the heuristic you proposed was "faulting address is close enough to the stack pointer", without any reference to a "top of stack" address.

File: runtime/backtrace.c [hunk 3]
```diff
@@ -136,34 +136,22 @@ CAMLprim value caml_get_exception_raw_backtrace(value unit)
...
-      saved_caml_backtrace_pos = BACKTRACE_BUFFER_SIZE;
-    }
-
-    memcpy(saved_caml_backtrace_buffer, caml_backtrace_buffer,
-           saved_caml_backtrace_pos * sizeof(backtrace_slot));
-
-    res = caml_alloc(saved_caml_backtrace_pos, 0);
-    for (i = 0; i < saved_caml_backtrace_pos; i++) {
-      Field(res, i) = Val_backtrace_slot(saved_caml_backtrace_buffer[i]);
+    /* We use caml_alloc_shr instead of caml_alloc, since caml_alloc might
```
Comment (xavierleroy) [inline runtime/backtrace.c]: All these changes in backtrace.c are unrelated to this PR, right?

**Review by xavierleroy — COMMENTED** [review]
Nice to see this work restart.  A couple of questions below.

File: runtime/signals_nat.c (hunk 2 above)
Comment (stedolan) [reply runtime/signals_nat.c]: > I'm afraid caml_top_of_stack is valid only for the initial thread and makes no sense for other threads.

I also thought so, but yesterday discovered that `st_stubs.c` contains code in `caml_thread_{save,restore}_runtime_state` to keep `caml_top_of_stack` updated and relevant to the current thread.

> I though the heuristic you proposed was "faulting address is close enough to the stack pointer", without any reference to a "top of stack" address.

That was indeed what I initially proposed. I'm happy to use that heuristic instead of caml_top_of_stack, but since it seems to be maintained for each thread I thought we may as well use it. Are there further accuracy problems with caml_top_of_stack?

File: runtime/backtrace.c (hunk 3 above)
Comment (stedolan) [reply runtime/backtrace.c]: They're in a seperate commit, but a related one. (It may be easier to read the PR commit-by-commit).

One of the changes to fix backtraces is to lower the stack probe amount from 32k to 4k, since a 32k stack probe is not reliable (on my machine, glibc pthreads uses a single 4k guard page, and allocates memory just afterwards for other purposes, so a 32k stack probe has false positives).

Only having 4k of stack guaranteed means that it is unwise to allocate 8k arrays on the stack, especially in functions that are likely to be called from a `Stack_overflow` handler. `caml_get_exception_raw_backtrace` was such a function, and the second commit in this PR fixes it.

**Comment by gasche** [comment]
846ce834e5dfc9e98c673bb752ca368a9dbffab3 (the change to `runtime/backtrace.c` to use `caml_alloc_shr` instead of `caml_alloc`, and get rid of a large on-stack buffer) looks like a very nice change on its own -- depending on how fast you think the present PR is going to get a decision, it may make sense to send it as a separate pre-PR.

**Comment by let-def** [comment]
@gasche @stedolan: Doesn't this risk to put a lot of short-lived values in the major heap? I can see this happening when `get_raw_backtrace` is used in conjunction with `reraise_with_backtrace`.

**Comment by gasche** [comment]
When I first thought about the change, my idea was that callstacks are typically (1) fairly large (often larger than Max_young_wosize?) and (2) likely to get stored later for auditing, and thus long-lived, or to come just before the program errors out, and thus we don't care.

(I had another paragraph here which I guess was confused about when backtraces are generated by the compiler when using exceptions for control-flow, and people are not using `raise_notrace`. I need to remind myself of how that actually works or let someone else comment on that part -- it could be indeed that this use-case generates a lot of backtraces and discards most of them immediately, but maybe not.)

**Comment by stedolan** [comment]
The situation with exceptions are used for control flow, where backtraces are unnecessarily generated and then discarded, is indeed annoyingly common. The importance of performance in this situation is why `caml_stash_backtrace` still has a static global buffer.

This change affects only Printexc.get_raw_backtrace (and a few other functions in `Printexc`), which use a seperate codepath from `caml_stash_backtrace`. I don't think this is used anywhere performance-sensitive.

**Comment by gasche** [comment]
Will the backtraces collected by statmemprof be affected? (Those are expected to follow the lifetime of the memory block they track, so we can expect that many of them will be short-lived.)

**Comment by stedolan** [comment]
No, statmemprof does not use this mechanism to collect callstacks. Statmemprof is interested in the current callstack, so it walks the stack to collect it. These functions concern the backtrace of the most recent exception, which is stored in a global buffer.

File: runtime/signals_nat.c (hunk 2 above)
Comment (stedolan) [reply runtime/signals_nat.c]: For reference: `caml_top_of_stack` was fixed by #996

**Comment by stedolan** [comment]
I changed this patch to remove the sharing of sigaltstacks. The reasoning why this was OK was based on considering only SIGSEGV; there are other signals and they share a stack.

File: runtime/signals_nat.c (hunk 1 above)
Comment (avsm) [reply runtime/signals_nat.c:198]: Thanks for the clarification, that sounds right to me. With a custom guard page limit it'll always work since there is no known usage of a page size < 4KB in the wild on any CPU architecture, which is the only way you could get a segfault without overrunning EXTRA_STACK.

File: runtime/signals_nat.c (hunk 2 above)
Comment (xavierleroy) [reply runtime/signals_nat.c]: You're right, I misremembered / mis-grepped the handling of `caml_top_of_stack`.  We need to maintain this variable in order to compute the stack usage part of GC stats, so why not use it, indeed. 

I think its value is reasonably accurate -- up to the size of the frame size of a small C runtime function, that's not much uncertainty.  The way `caml_top_of_stack` is initialized is very much undefined C behavior, but that hasn't been a problem so far, and could be changed later.

**Comment by xavierleroy** [comment]
> This change affects only Printexc.get_raw_backtrace (and a few other functions in Printexc), which use a seperate codepath from caml_stash_backtrace.

This is true.

> I don't think this is used anywhere performance-sensitive.

This is perhaps less true these days.  For example, `Fun.protect` calls `Printexc.get_raw_backtrace` if an exception occurs.  I would still hope that `Fun.protect` is not used in conjunction with functions that raise often, but who knows...

**Comment by gasche** [comment]
I looked for usage of Misc.try_finally (a parent of Fun.protect)
in the compiler codebase, I have the impression that all
use-cases concern blocks where the exception path is indeed
exceptional. It is also used at a fairly coarse
granularity (typically, the cleanup action is to remove
a temporary file), so I think additional costs would be
amortized.

I haven't seen a use of Fun.protect around exceptions used for
control-flow, but that seems entirely possible to me. (For example,
a SAT solver where the "restart" action, every N decision steps,
is implemented by raising an exception, may use Fun.protect to unwind
the decision stack.) I think it would make sense to also offer a
Fun.protect_notrace function, cousin of raise_notrace, that does
not capture the call stack and uses a reraise_notrace.

Let me know if you like the idea and I'll shoot a separate PR or issue report.

**Comment by stedolan** [comment]
> I think it would make sense to also offer a
Fun.protect_notrace function, cousin of raise_notrace, that does
not capture the call stack and uses a reraise_notrace.

Currently (both before and after this patch), if `raise_notrace` is used, then `Printexc.get_raw_backtrace` is very fast (it does not allocate at all, since there is no backtrace). So I don't think this function is necessary. Effectively, `Fun.protect` does a `reraise_notrace` if the original exception came from `raise_notrace`, which I think is the right behaviour.

**Comment by gasche** [comment]
Ah, indeed. Sorry for the internal confusedness.

**Comment by gadmm** [comment]
> I looked for usage of Misc.try_finally (a parent of Fun.protect)
> in the compiler codebase, I have the impression that all
> use-cases concern blocks where the exception path is indeed
> exceptional. It is also used at a fairly coarse
> granularity (typically, the cleanup action is to remove
> a temporary file), so I think additional costs would be
> amortized.

In the following review of uses of `try_finally` in the compiler: https://github.com/ocaml/ocaml/pull/2118#issuecomment-441467204, I have found several cases where it is used for expressing temporary mutation, where the `finally` just rolls back some state. This is an idiomatic use of unwind-protect, and is a use-case that can occur in algorithmically-intensive situations (e.g. if exceptions are used for control flow). It seems incorrect to me to assume in general that resources are few and disposal is expensive (amortizes the cost of recording the stack trace).

> I haven't seen a use of Fun.protect around exceptions used for
> control-flow, but that seems entirely possible to me. (For example,
> a SAT solver where the "restart" action, every N decision steps,
> is implemented by raising an exception, may use Fun.protect to unwind
> the decision stack.)

Yes, this is a valid use-case, actually I have found 5 of them in the aforementioned audit. Control flow is the intended usage of `Fun.protect`, when distinguished from `bracket`-style resource management that would do masking for async exceptions. This use-case was in fact the reason put forward in favour of including `Fun.protect` with its current type, whatever the fate of asynchronous exceptions would be.

Out of curiosity, how expensive is it to call `Printexc.get_raw_backtrace` immediately after `Printexc.raise_with_backtrace`, for instance in the case where the user has nested `Fun.protect`s?

**Comment by stedolan** [comment]
This discussion got side-tracked by some questions about whether allocating backtraces on the major heap would have a noticeable performance impact.

I've just rebased this patch to trunk. On trunk, since #8691, plain `caml_alloc` should not trigger callbacks. So, the new version of this patch just calls `caml_alloc` (which allocates on the minor or major heap as appropriate).

However, when rebasing this I noticed a bug in #8691, in that it does not always prevent `caml_alloc` from triggering callbacks. This bug is fixed in #8897, so we shouldn't merge this patch until after #8897.

**Comment by stedolan** [comment]
#8897 is merged, and this branch has just passed precheck (300), so I think this is ready to merge.

**Comment by gadmm** [comment]
The second commit about removing the stack-allocated buffer looks good to me. (I would happily help with a review of the first commit but the technical details still fly over my head.)

File: runtime/backtrace.c [hunk 4]
```diff
@@ -132,34 +132,17 @@ CAMLprim value caml_get_exception_raw_backtrace(value unit)
...
-           saved_backtrace_pos * sizeof(backtrace_slot));
+    intnat i, len = Caml_state->backtrace_pos;
 
-    res = caml_alloc(saved_backtrace_pos, 0);
-    for (i = 0; i < saved_backtrace_pos; i++) {
-      Field(res, i) = Val_backtrace_slot(saved_backtrace_buffer[i]);
-    }
+    res = caml_alloc(len, 0);
+    for (i = 0; i < len; i++)
+      Field(res, i) = Val_backtrace_slot(Caml_state->backtrace_buffer[i]);
```
Comment (gadmm) [inline runtime/backtrace.c:145]: Maybe I am still learning the backtrace mechanism, but how are you sure that `i < BACKTRACE_BUFFER_SIZE`?

File: Changes [hunk 5]
```diff
@@ -73,6 +73,10 @@ Working version
   (Stephen Dolan, Xavier Leroy and David Allsopp,
    review by Xavier Leroy and Gabriel Scherer)
 
+- #8670: Fix stack overflow detection with systhreads
+  (Stephen Dolan, review by Xavier Leroy, Anil Madhavapeddy, Gabriel Scherer,
+   Frédéric Bour and Guillaume Munch-Maccagnoni)
```
Comment (gadmm) [inline Changes:78]: Sorry again for only commenting on the second commit which is unrelated to the topic of this PR. Please remove me from the Changes file, this is not deserved.

File: runtime/backtrace.c (hunk 4 above)
Comment (stedolan) [reply runtime/backtrace.c:145]: The prefix of the backtrace buffer that's in use is the index range `0` to `Caml_state->backtrace_pos - 1`, and so the invariant is maintained that `Caml_state->backtrace_pos` (i.e. `len`) is always `<= BACKTRACE_BUFFER_SIZE` (See `caml_stash_backtrace` in `backtrace_nat.c`).

File: Changes (hunk 5 above)
Comment (stedolan) [reply Changes:78]: I think my comments above were phrased poorly, apologies. Let me rephrase: the original version of this code introduced a potential performance issue with Printexc.get_raw_backtrace. I'd thought this couldn't matter, but you and others pointed out that it could.

Happily, the issue went away when #8691 was merged, so the discussion about whether or not the issue mattered became moot: now there is no problem simply calling caml_alloc.

Still, the original discussion was a useful review of the code at the time, so I think your review helped and I'd lean towards leaving your name there. (Of course, I'll delete it if you'd still prefer that).

Comment (gadmm) [reply Changes:78]: I was expressing my discomfort of being associated with the original topic of the PR, which is quite a major improvement but for which I felt I did nothing. So, thank you for the clarification!

File: runtime/backtrace.c (hunk 4 above)
Comment (gadmm) [reply runtime/backtrace.c:145]: Essentially, `if (saved_backtrace_pos > BACKTRACE_BUFFER_SIZE)` was in fact already redundant in the previous code. I even found what you wrote documented in `backtrace.h`, so all is clear now, thanks.

**Comment by stedolan** [comment]
I'd quite like to get this into 4.10. Are there any remaining issues?

**Review by xavierleroy — APPROVED** [review]
I had a second look at the code and at the discussion.  I agree it's good to go.

**Comment by stedolan** [comment]
Thanks!

**Comment by gadmm** [comment]
I hope it's the right place to report it: I have a bunch of errors like the following one when I do `make depend` inside the `runtime` directory:
```sh
gcc -MM -D_FILE_OFFSET_BITS=64 -D_REENTRANT -DCAML_NAME_SPACE   *.c | \
  sed -e 's/\([^.]*\)\.o/\1_b.$(O)/' > .depend
signals_nat.c:192:2: error: #error "CONTEXT_SP is required if HAS_STACK_OVERFLOW_DETECTION is defined"
 #error "CONTEXT_SP is required if HAS_STACK_OVERFLOW_DETECTION is defined"
  ^~~~~
```
